llarp/net/ip_packet.hpp

```diff
@@ -4,324 +4,261 @@
+    ICMP6 = 0x3A,
+    OSFP = 0x59,
+    PGM = 0x71,
+  };
+
+  /// get string representation of this protocol
+  /// throws std::invalid_argument if we don't know the name of this ip protocol
+  std::string
+  IPProtocolName(IPProtocol proto);
+
+  /// parse a string to an ip protocol
+  /// throws std::invalid_argument if cannot be parsed
+  IPProtocol
+  ParseIPProtocol(std::string data);
+
+  /// an Packet
+  struct IPPacket
+  {
+    static constexpr size_t MaxSize = 1500;
+    llarp_time_t timestamp;
+    size_t sz;
+    byte_t buf[MaxSize];
 
-      static IPPacket
-      UDP(nuint32_t srcaddr,
-          nuint16_t srcport,
-          nuint32_t dstaddr,
-          nuint16_t dstport,
-          const llarp_buffer_t& data);
+    static IPPacket
+    UDP(nuint32_t srcaddr,
+        nuint16_t srcport,
+        nuint32_t dstaddr,
+        nuint16_t dstport,
+        const llarp_buffer_t& data);
 
-      ManagedBuffer
-      Buffer();
+    ManagedBuffer
+    Buffer();
 
-      ManagedBuffer
-      ConstBuffer() const;
+    ManagedBuffer
+    ConstBuffer() const;
 
-      bool
-      Load(const llarp_buffer_t& buf);
+    bool
+    Load(const llarp_buffer_t& buf);
 
-      struct GetTime
-      {
-        llarp_time_t
-        operator()(const IPPacket& pkt) const
-        {
-          return pkt.timestamp;
-        }
-      };
-
-      struct PutTime
-      {
-        EventLoop_ptr loop;
-        PutTime(EventLoop_ptr evloop) : loop(std::move(evloop))
-        {}
-        void
-        operator()(IPPacket& pkt) const
-        {
-          pkt.timestamp = loop->time_now();
-        }
-      };
-
-      struct GetNow
-      {
-        EventLoop_ptr loop;
-        GetNow(EventLoop_ptr evloop) : loop(std::move(evloop))
-        {}
-        llarp_time_t
-        operator()() const
-        {
-          return loop->time_now();
-        }
-      };
-
-      struct CompareSize
-      {
-        bool
-        operator()(const IPPacket& left, const IPPacket& right)
-        {
-          return left.sz < right.sz;
-        }
-      };
-
-      struct CompareOrder
-      {
-        bool
-        operator()(const IPPacket& left, const IPPacket& right)
-        {
-          return left.timestamp < right.timestamp;
-        }
-      };
-
-      inline ip_header*
-      Header()
+    struct CompareSize
+    {
+      bool
+      operator()(const IPPacket& left, const IPPacket& right)
       {
-        return (ip_header*)&buf[0];
+        return left.sz < right.sz;
       }
+    };
 
-      inline const ip_header*
-      Header() const
+    struct CompareOrder
+    {
+      bool
+      operator()(const IPPacket& left, const IPPacket& right)
       {
-        return (ip_header*)&buf[0];
+        return left.timestamp < right.timestamp;
       }
+    };
 
-      inline ipv6_header*
-      HeaderV6()
-      {
-        return (ipv6_header*)&buf[0];
-      }
+    inline ip_header*
+    Header()
+    {
+      return reinterpret_cast<ip_header*>(&buf[0]);
```

Comment:
`buf` should probably be declared: `alignas(ip_header) byte_t buf[MaxSize];` to make these reinterpret_cast's valid.

Reply:
resolved

Reply:
(Given that buf follows a size_t member it is *probably* already properly aligned, but still worth being explicit about the alignment requirement).